### PR TITLE
Circuit breaker for individual log messages

### PR DESCRIPTION
**Pull request by slacAWallace**
@n-wbrown @ghalym

### CODE REVIEW

**Comment by klauer** [comment]
Does this work? I'd imagine you'd want to stop the messages from ever going to the event system, not ignore them once they're already created

**Comment by slacAWallace** [comment]
You mean at where the log message is generated?

**Comment by klauer** [comment]
Yeah. If it gets passed to the event system, you'll see the message in TwinCAT. 

Stopping the log handler (which listens for events from the event system) will cause messages to get backed up in the PLC buffers. Of course, they will also not be sent off to logstash/grafana here.

**Comment by slacAWallace** [comment]
@n-wbrown @klauer tell me what you think of this. I still need to pytmc it, but I think it solves both the very problematic single log-message and the trickle trip cases.

**Comment by slacAWallace** [comment]
@klauer @n-wbrown poke

**Comment by slacAWallace** [comment]
OK now it actually does what I was intending at least for the local trip. The global trickle trip still needs a test to target the condition I am trying to address, but at least the local trip will stop really spammy loggers from DoSing a PLC. I would appreciate feedback on how intuitive the circuit breaker settings are.

File: LCLSGeneral/LCLSGeneral/POUs/Logger/GVL_Logger.TcGVL [hunk 1]
```diff
@@ -28,6 +28,16 @@ VAR_GLOBAL CONSTANT
 	iLogPort		:	UINT		:= 54321;
     
     sIpTidbit : STRING(6) := '172.21';
+   
+    // Log message circuit breaker configuration 
+    
+    // Initialization constants
+    nLocalTripThreshold : TIME := T#1ms; // Minimum time between logs
+    nMinTimeViolationAcceptable : UINT := 5; // Maximum number of times the minimum time above can be violated before a trip
+    nLocalTrickleTripThreshold : TIME := T#100ms; // Trickle trip, activated by global threshold (units: 100ns)
```
Comment (klauer) [inline LCLSGeneral/LCLSGeneral/POUs/Logger/GVL_Logger.TcGVL:37]:
```suggestion
    nLocalTrickleTripThreshold : TIME := T#100ms; // Trickle trip, activated by global threshold
```

File: LCLSGeneral/LCLSGeneral/POUs/Logger/GVL_Logger.TcGVL [hunk 2]
```diff
@@ -28,6 +28,16 @@ VAR_GLOBAL CONSTANT
 	iLogPort		:	UINT		:= 54321;
     
     sIpTidbit : STRING(6) := '172.21';
+   
+    // Log message circuit breaker configuration 
+    
+    // Initialization constants
+    nLocalTripThreshold : TIME := T#1ms; // Minimum time between logs
```
Comment (klauer) [inline LCLSGeneral/LCLSGeneral/POUs/Logger/GVL_Logger.TcGVL:35]:
```suggestion
    nLocalTripThreshold : TIME := T#1ms; // Minimum time between log messages
```

File: LCLSGeneral/LCLSGeneral/POUs/Logger/GVL_Logger.TcGVL [hunk 3]
```diff
@@ -28,6 +28,16 @@ VAR_GLOBAL CONSTANT
...
+    nMinTimeViolationAcceptable : UINT := 5; // Maximum number of times the minimum time above can be violated before a trip
```
Comment (klauer) [inline LCLSGeneral/LCLSGeneral/POUs/Logger/GVL_Logger.TcGVL:36]: Maybe?
```suggestion
    nMinTimeViolationAcceptable : UINT := 5; // Trip if `nLocalTripThreshold` exceeded `nMinTimeViolationAcceptable` times
```

**Comment by klauer** [comment]
Given that you accepted my suggestions it looks like I was right in the interpretation of them. In that case, I think they're reasonably clear once you know the overall intent of the circuit breaker.

**Comment by slacAWallace** [comment]
Severity	Code	Description	Project	File	Line	Suppression State
Error		8/5/2020 3:19:40 PM 097 ms | 'Task 3' (350): | ======================================				
Error		8/5/2020 3:19:38 PM 497 ms | 'Task 3' (350): | ==========TESTS FINISHED RUNNING==========				
Error		8/5/2020 3:19:38 PM 817 ms | 'Task 3' (350): | Test suites: 1				
Error		8/5/2020 3:19:39 PM 137 ms | 'Task 3' (350): | Tests: 3				
Error		8/5/2020 3:19:39 PM 457 ms | 'Task 3' (350): | Successful tests: 3				
Error		8/5/2020 3:19:39 PM 777 ms | 'Task 3' (350): | Failed tests: 0

**Comment by slacAWallace** [comment]
@klauer if this looks OK i am going to pull it in now. It would help us today.

**Comment by klauer** [comment]
👏 Let's see how it goes!

**Pull request merged**